modules/google-analytics/wp-google-analytics.php

```diff
@@ -22,312 +22,55 @@
 	 */
-	private function _get_url( $track ) {
-		$site_url = ( is_ssl() ? 'https://':'http://' ) . sanitize_text_field( wp_unslash( $_SERVER['HTTP_HOST'] ) ); // Input var okay.
-		foreach ( $track as $k => $value ) {
-			if ( strpos( strtolower( $value ), strtolower( $site_url ) ) === 0 ) {
-				$track[ $k ] = substr( $track[ $k ], strlen( $site_url ) );
-			}
-			if ( 'data' === $k ) {
-				$track[ $k ] = preg_replace( '/^https?:\/\/|^\/+/i', '', $track[ $k ] );
-			}
-
-			// This way we don't lose search data.
-			if ( 'data' === $k && 'search' === $track['code'] ) {
-				$track[ $k ] = rawurlencode( $track[ $k ] );
-			} else {
-				$track[ $k ] = preg_replace( '/[^a-z0-9\.\/\+\?=-]+/i', '_', $track[ $k ] );
-			}
-
-			$track[ $k ] = trim( $track[ $k ], '_' );
-		}
-		$char = ( strpos( $track['data'], '?' ) === false ) ? '?' : '&amp;';
-		return str_replace( "'", "\'", "/{$track['code']}/{$track['data']}{$char}referer=" . rawurlencode( isset( $_SERVER['HTTP_REFERER'] ) ? $_SERVER['HTTP_REFERER'] : '' ) ); // Input var okay.
-	}
+	static $instance = false;
 
 	/**
-	 * Maybe output or return, depending on the context
+	 * @var Static property to hold concrete analytics impl that does the work (universal or legacy)
 	 */
-	private function _output_or_return( $val, $maybe ) {
-		if ( $maybe ) {
-			echo $val . "\r\n";
-		} else {
-			return $val;
-		}
-	}
+	static $analytics = false;
 
 	/**
-	 * This injects the Google Analytics code into the footer of the page.
+	 * This is our constructor, which is private to force the use of get_instance()
 	 *
-	 * @param bool[optional] $output - defaults to true, false returns but does NOT echo the code.
+	 * @return void
 	 */
-	public function insert_code( $output = true ) {
-		// If $output is not a boolean false, set it to true (default).
-		$output = ( false !== $output);
-
-		$tracking_id = $this->_get_tracking_code();
-		if ( empty( $tracking_id ) ) {
-			return $this->_output_or_return( '<!-- Your Google Analytics Plugin is missing the tracking ID -->', $output );
-		}
-
-		// If we're in the admin_area, return without inserting code.
-		if ( is_admin() ) {
-			return $this->_output_or_return( '<!-- Your Google Analytics Plugin is set to ignore Admin area -->', $output );
-		}
-
-		$custom_vars = array(
-			"_gaq.push(['_setAccount', '{$tracking_id}']);",
-		);
-
-		$track = array();
-		if ( is_404() ) {
-			// This is a 404 and we are supposed to track them.
-			$custom_vars[] = "_gaq.push(['_trackEvent', '404', document.location.href, document.referrer]);";
-		} elseif ( is_search() ) {
-			// Set track for searches, if it's a search, and we are supposed to.
-			$track['data'] = sanitize_text_field( wp_unslash( $_REQUEST['s'] ) ); // Input var okay.
-			$track['code'] = 'search';
+	private function __construct() {
+		if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
+			error_log( print_r( Jetpack_Google_Analytics_Options::debug_dump(), true ) );
```

Comment:
Is this `error_log` needed?, I'm seeing it's only dumping empty arrays even when nothing remarkable happens.

Reply:
Its just for when WP_DEBUG is active. It isn't as useful as I'd thought - I think I'll just delete the call and the method.

Reply:
Removed in https://github.com/Automattic/jetpack/pull/8182/commits/dca827885f68c2d00f741f6778b592c16f6b09ed